assembl/fabfile.py

```diff
@@ -1472,19 +1610,114 @@ def generate_dh_group():
+        with settings(user=env.webmaster_user, passwords=fabpass_for(env.webmaster_user)):
+            put(file, "/etc/nginx/sites-available/%s" % (env.public_hostname))
+            status = sudo('nginx -t', warn_only=True)
+            if status.failed:
+                run('rm -f /etc/nginx/sites-available/%s' % env.public_hostname)
+                run('mv /etc/nginx/sites-available/%(host_name)s.bak /etc/nginx/sites-available/%(host_name)s' % {'host_name': env.public_hostname})
+                raise RuntimeError("The NGINX configuration was improperly set up! Please rerun setup_nginx_file with debug enabled to view the config")
+            else:
+                run("rm -f /etc/nginx/sites-enabled/%s" % (env.public_hostname))
+                run("ln -s /etc/nginx/sites-available/%(host_name)s /etc/nginx/sites-enabled/%(host_name)s" % {'host_name': env.public_hostname})
+                run("sudo /etc/init.d/nginx restart")
+    finally:
+        if debug:
+            os.unlink(file)
+
+
+def get_s3_file(bucket, key, destination=None):
+    import boto3
+    s3 = boto3.client('s3')
+    try:
+        ob = s3.get_object(Bucket=bucket, Key=key)
+        content = ob['Body'].read()
+        if destination:
+            with open(destination, 'w') as f:
+                f.write(content)
+        return content
+    except s3.exceptions.NoSuchKey:
+        return None
+
+
+@task
+def get_aws_localrc():
+    assert running_locally()
+    import requests
+    r = requests.get('http://169.254.169.254/latest/meta-data/iam/info')
+    assert r.ok
+    account = r.json()['InstanceProfileArn'].split(':')[4]
+    # This introduces a convention: local.rc files
+    # for a given amazon account will be stored in
+    # s3://bluenove-assembl-configurations/local_{account_id}.rc
+    content = get_s3_file(
+        'bluenove-assembl-configurations',
+        'local_%s.rc' % account,
+        env.projectpath + '/local.rc')
+    extends_re = re.compile(r'\b_extends\s*=\s*(\w+\.rc)')
+    match = extends_re.search(content)
+    while match:
+        key = match.group(1)
+        ex_content = get_s3_file('bluenove-assembl-configurations', key)
+        if not ex_content:
+            break
+        with open(key, 'w') as f:
+            f.write(ex_content)
+        match = extends_re.search(ex_content)
+    if not content:
+        raise RuntimeError("local_%s.rc was not defined in S3" % account)
+
+
+@task
+def setup_aws_default_region():
+    assert running_locally()
+    import requests
+    r = requests.get('http://169.254.169.254/latest/meta-data/placement/availability-zone')
+    assert r.ok
+    region = r.content[:-1]
+    run('aws configure set region ' + region)
+
+
+@task
+def aws_instance_startup():
+    """Operations to startup a fresh aws instance from an assembl AMI"""
+    if not env.projectpath:
+        # Assumption: Started from project directory
+        env.projectpath = os.cwd()
```

Comment:
`os.cwd()` doesn't exist! It should be `os.getcwd()`

Reply:
Lol, good catch. This code was never tested, and in fact the logic was recreated by invoke.